staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_discovery.go

```diff
@@ -0,0 +1,617 @@
+		} else {
+			path = "/apis/" + gv.Group + "/" + gv.Version
+		}
+
+		req, err := http.NewRequest("GET", path, nil)
+		if err != nil {
+			// NewRequest should not fail, but if it does for some reason,
+			// log it and continue
+			return cached, fmt.Errorf("failed to create http.Request: %v", err)
+		}
+
+		// Apply aggregator user to request
+		req = req.WithContext(
+			request.WithUser(
+				req.Context(), &user.DefaultInfo{Name: "system:kube-aggregator"}))
+
+		// req.Header.Add("Accept", runtime.ContentTypeProtobuf)
+		req.Header.Add("Accept", runtime.ContentTypeJSON)
+
+		if exists && len(cached.etag) > 0 {
+			req.Header.Add("If-None-Match", cached.etag)
+		}
+
+		writer := newInMemoryResponseWriter()
+		handler.ServeHTTP(writer, req)
+
+		if writer.respCode != http.StatusOK {
+			return cached, fmt.Errorf("failed to download discovery for %s: %v", path, writer.String())
+		}
+
+		parsed := &metav1.APIResourceList{}
+		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
+			return cached, err
+		}
+
+		// Create a discomap with single group-version
+		resources, err := endpoints.ConvertGroupVersionIntoToDiscovery(parsed.APIResources)
+		if err != nil {
+			return cached, err
+		}
+
+		discoMap := map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery{
+			// Convert old-style APIGroupList to new information
+			gv: {
+				Version:   gv.Version,
+				Resources: resources,
+			},
+		}
+
+		cached = cachedResult{
+			discovery:   discoMap,
+			lastUpdated: now,
+		}
+
+		// Don't bother saving result, there is no ETag support on this endpoint
+		dm.cachedResults[info.service] = cached
+
+		return cached, nil
+
+	case http.StatusOK:
+		parsed := &apidiscoveryv2beta1.APIGroupDiscoveryList{}
+		if err := runtime.DecodeInto(scheme.Codecs.UniversalDecoder(), writer.data, parsed); err != nil {
+			return cached, err
+		}
+		klog.Infof("DiscoveryManager: Successfully downloaded discovery for %s", info.service.String())
+
+		// Convert discovery info into a map for convenient lookup later
+		discoMap := map[metav1.GroupVersion]apidiscoveryv2beta1.APIVersionDiscovery{}
+		for _, g := range parsed.Items {
+			for _, v := range g.Versions {
+				discoMap[metav1.GroupVersion{Group: g.Name, Version: v.Version}] = v
+			}
+		}
+
+		dm.resultsLock.Lock()
+		defer dm.resultsLock.Unlock()
+
+		// Save cached result
+		cached = cachedResult{
+			discovery:   discoMap,
+			etag:        writer.Header().Get("Etag"),
+			lastUpdated: now,
+		}
+		dm.cachedResults[info.service] = cached
+		return cached, nil
+
+	default:
+		dm.resultsLock.Lock()
+		defer dm.resultsLock.Unlock()
+
+		// Unhandled response. Mark information as stale.
+		// Try again later.
+		//!TODO: After a few tries, just wipe it out?
+		// 	or after certain time?
+		if !cached.stale {
+			cached = cachedResult{
+				discovery:   cached.discovery,
+				etag:        cached.etag,
+				lastUpdated: now,
+				stale:       true,
```

Comment:
is this the only spot setting stale?  What happens if we try to refresh several times after `lastMarkedDirty`is set?

Reply:
you were right, this was not sufficient to determine staleness.

I've refactored the logic more linearly so hopefully it is clear how stale is determined: https://github.com/kubernetes/kubernetes/pull/113171/commits/b5749b258bc959b30774e2ecb6d41fca4270e050